internal/registration/runner.go

```diff
@@ -0,0 +1,195 @@
+		err = errors.New("registration: missing public key")
+
+	case publicKey != "" && tokenPath == "":
+		err = errors.New("registration: missing token path")
+	}
+
+	return runner, err
+}
+
+// CheckToken loads and verifies the configured token, returning an error when
+// the file exists but cannot be verified.
+func (r *Runner) CheckToken() error {
+	data, errFile := os.ReadFile(r.tokenPath)
+	key := func(*jwt.Token) (any, error) { return r.publicKey, nil }
+
+	// Assume [jwt] and [os] functions could do something unexpected; use defer
+	// to safely write to the token.
+	r.token.Lock()
+	defer r.token.Unlock()
+
+	_, errToken := jwt.ParseWithClaims(string(data), &r.token, key,
+		jwt.WithExpirationRequired(),
+		jwt.WithValidMethods([]string{"RS256"}),
+	)
+
+	// The error from [os.ReadFile] indicates whether a token file exists.
+	r.token.Exists = !os.IsNotExist(errFile)
+
+	// Reset most claims if there is any problem loading, parsing, validating, or
+	// verifying the token file.
+	if errFile != nil || errToken != nil {
+		r.token.RegisteredClaims = jwt.RegisteredClaims{}
+	}
+
+	switch {
+	case !r.enabled || !r.token.Exists:
+		return nil
+	case errFile != nil:
+		return errFile
+	default:
+		return errToken
+	}
+}
+
+func (r *Runner) state() (failed, required bool) {
+	// Assume [time] functions could do something unexpected; use defer to safely
+	// read the token.
+	r.token.RLock()
+	defer r.token.RUnlock()
+
+	failed = r.token.Exists && r.token.ExpiresAt == nil
+	required = r.enabled &&
+		(!r.token.Exists || failed || r.token.ExpiresAt.Before(time.Now()))
+	return
+}
+
+// Required returns true when registration is required but the token is missing or invalid.
+func (r *Runner) Required(
+	recorder record.EventRecorder, object client.Object, conditions *[]metav1.Condition,
+) bool {
+	failed, required := r.state()
+
+	if r.enabled && failed {
+		emitFailedWarning(recorder, object)
+	}
+
+	if !required && conditions != nil {
+		before := len(*conditions)
+		meta.RemoveStatusCondition(conditions, v1beta1.Registered)
+		meta.RemoveStatusCondition(conditions, "RegistrationRequired")
+		meta.RemoveStatusCondition(conditions, "TokenRequired")
+		found := len(*conditions) != before
+
+		if r.enabled && found {
+			emitVerifiedEvent(recorder, object)
+		}
+	}
+
+	return required
+}
+
+// NeedLeaderElection returns true so that r runs only on the single
+// [manager.Manager] that is elected leader in the Kubernetes namespace.
+func (r *Runner) NeedLeaderElection() bool { return true }
+
+// Start watches for a mounted registration token when enabled. It blocks
+// until ctx is cancelled.
+func (r *Runner) Start(ctx context.Context) error {
+	var ticks <-chan time.Time
+
+	if r.enabled {
+		ticker := time.NewTicker(r.refresh)
+		defer ticker.Stop()
+		ticks = ticker.C
+	}
+
+	log := logging.FromContext(ctx).WithValues("controller", "registration")
+
+	for {
+		select {
+		case <-ticks:
+			_, before := r.state()
```

Comment:
:question: How does the Runner's state persist across a restart? Or does it have to?

Reply:
The state is loaded once during startup before the Manager calls this function:

https://github.com/CrunchyData/postgres-operator/blob/3315e0cb3bbb70b12bf635758c57b1b0ac1bdad3/cmd/postgres-operator/main.go#L105